mitodl/eslint-config
# Support Typescript, Jest

#### Pre-Flight checklist


#### What are the relevant tickets?
Closes #14 
Closes #15 

#### What's this PR do?
This PR:
- updates our linting config to support Typescript and Jest.
- resurrects support for Flow *Flow support was dropped in #12. I've changed the setup a bit so that we can support Flow and Typescript. This is not to say that we should use flow in new projects. But continuing flow is not particularly hard, and makes it easier for flow+typescript to live side by side, which could be useful as we migrate a project from flow to TS. As a concrete example, https://github.com/mitodl/open-discussions/pull/3763 updates Open Discussions + Infinite Corridor to use this branch, a repo in which flow and TS coexist.*
- Switches the package manager from npm to yarn 3, since that's what we've been using lately.

In order to support Jest or Mocha, Flow or TS, I restructured the package a lit bit. Previously there was a single export. Now there are a few:
```
eslint-config-mitodl            <--- the "base" config, which includes TS support but can also lint regular JS
eslint-config-mitodl/jest
eslint-config-mitodl/mocha
eslint-config-mitodl/flow
```

Example: TS + jest [Infinite Corridor's new config](https://github.com/mitodl/open-discussions/pull/3763/files#diff-e2954b558f2aa82baff0e30964490d12942e0e251c1aa56c3294de6ec67b7cf5R2-R3): 
```js
module.exports = {
  extends:   ["eslint-config-mitodl", "eslint-config-mitodl/jest", "plugin:testing-library/react"],
  plugins: ["testing-library"],
}
```

Example: Flow + Mocha [Open Discussions' new config](https://github.com/mitodl/open-discussions/pull/3763/files#diff-e2954b558f2aa82baff0e30964490d12942e0e251c1aa56c3294de6ec67b7cf5R2-R3): 
```js
module.exports = {
  "extends": [
    "eslint-config-mitodl",
    "eslint-config-mitodl/flow",
    "eslint-config-mitodl/mocha"
  ],
  // a few other overrides
}
```

#### How should this be manually tested?
1. Clone this repo if you don't have it locally, and `yarn install`.
2. Run `yarn run test`. It should succeed.
3. Make a breaking edit to our linting config, like change [this line](https://github.com/mitodl/eslint-config/pull/16/files#diff-de1741a49b319b1d44b86438ad991cc0e33a37c577461f4890a96505a19f06c6R4) to use `testz` instead of `test` as the file matching pattern. Run `yarn run test` again. It should fail.
4.  View 
    - https://github.com/mitodl/open-discussions/pull/3763, which updates Open Discussions + Infinite Corridor to use the new config
    - https://github.com/mitodl/ocw-studio/pull/1536, which updates OCW Studio to use the new config 

    Observe that the linting changes are very minor. The goal of this PR is not to make significant changes to our linting config, but rather to update dependencies + make TS & Jest support easier.
5. If you want to be extra careful... clone one of the branches above, and do something that should cause a linting error, and run the corresponding linting command. You should get an error.
